src/sentry/api/base.py

```diff
@@ -402,6 +405,148 @@ def paginate(
+                f"`{wrapped.__name__}_v{version}`."
+            )
+
+        return _VersionedMethodBody(http_method, version, wrapped)
+
+    return decorator
+
+
+class _MethodVersionTable:
+    """A table of all handler methods on a versioned endpoint."""
+
+    def __init__(self, endpoint_cls) -> None:
+        self.table: Mapping[str, Mapping[int, _VersionedMethodBody]] = self._build(endpoint_cls)
+
+    def _build(self, endpoint_cls) -> Mapping[str, Mapping[int, _VersionedMethodBody]]:
+        table = defaultdict(dict)
+        for attr_name in dir(endpoint_cls):
+            attr_obj = getattr(endpoint_cls, attr_name)
+            if isinstance(attr_obj, _VersionedMethodBody):
+                method_table = table[attr_obj.http_method]
+                if attr_obj.version in method_table:
+                    raise Exception(
+                        "Multiple methods mapped to "
+                        f"[method={attr_obj.http_method!r}, version={attr_obj.version!r}] "
+                        f"in {endpoint_cls!r}"
+                    )
+                table[attr_obj.http_method][attr_obj.version] = attr_obj
+        return table
+
+    def supports_http_method(self, http_method: str) -> bool:
+        """Check whether the endpoint supports an HTTP method.
+
+        This determines whether a 405 response status is appropriate.
+        """
+        return bool(self.table[http_method])
+
+    def get_method_version(self, http_method: str, version: int) -> Optional[_VersionedMethodBody]:
+        """Look up a method version, or None if it doesn't exist."""
+        return self.table[http_method].get(version)
+
+
+class VersionedEndpoint(Endpoint):
+    """An endpoint with that takes requests for multiple versions.
+
+    Do not implement the handlers for HTTP methods (such as `get`) as you would
+    in a regular Endpoint class. Instead, decorate the handler methods with
+    `@method_version`.
+    """
+
+    _method_version_table = None
+
+    @classmethod
+    def _get_method_versions(cls) -> _MethodVersionTable:
+        table = cls._method_version_table  # for thread safety
+        if table is None:
+            cls._method_version_table = table = _MethodVersionTable(cls)
+        return table
+
+    def _dispatch_to_version(self, http_method: str, request, *args, **kwargs):
+        if http_method not in self.http_method_names:
+            raise ValueError(
+                f"Method name ({http_method!r}) must be one of: {self.http_method_names!r} "
+                "(don't call _dispatch_to_version from VersionedEndpoint subclasses)"
+            )
+
+        method_table = _MethodVersionTable(type(self))
+        if not method_table.supports_http_method(http_method):
+            return self.http_method_not_allowed(request)
+        version = 0 if request.version is None else int(request.version)
+        method_version = method_table.get_method_version(http_method, version)
+        if method_version is None:
```

Comment:
you can hook into a subclass definition using `__init_subclass__` which gets called at AST parse time. There you can verify the methods of the class to not have any erroneous ones.

Reply:
Thanks for the tip; see a98edddb624d786a4e79007bcaf518ed30095019. Though this `method_version is None` check reflects a client-side error (they requested a version that doesn't exist), so a 400-series response is appropriate.